### DISCUSSION TITLE
Lag and window features for exegenous variables.

### DISCUSSION THREAD

**Discussion by borisvettorato** [JoaquinAmatRodrigo/skforecast#856] [Q&A]
Is it possible to implement the lag and window features for the exogenous features in the forecaster too? I am trying to use the forecaster in an optimizer by changing one of the exogenous features and it would make my coding easier if the forecaster itself could adjust all the window and lag features of this exogenous feature.

**Comment by JoaquinAmatRodrigo** [comment]
Hi @borisvettorato,
In the current version of skforecast, all transformations of the exogenous variables should be done using the `transformer_exog` argument, otherwise outside the forecaster.

It may be possible to automate the creation of lag and window features for exogenous variables inside the forecaster using the [LagFeatures](https://feature-engine.trainindata.com/en/1.8.x/api_doc/timeseries/forecasting/LagFeatures.html) and [WindowFeatures](https://feature-engine.trainindata.com/en/1.8.x/api_doc/timeseries/forecasting/WindowFeatures.html#feature_engine.timeseries.forecasting.WindowFeatures) transformers available in feature-engine (@solegalli)

Let us know the result if you try it.

**Comment by JoaquinAmatRodrigo** [comment]
```python

# Libraries
# ==============================================================================
from sklearn.pipeline import make_pipeline
from feature_engine.timeseries.forecasting import WindowFeatures
from feature_engine.timeseries.forecasting import LagFeatures
from skforecast.datasets import fetch_dataset

# Downloading data
# ==============================================================================
data = fetch_dataset(name='bike_sharing', raw=False)
data = data.loc[:, ['users', 'holiday', 'temp', 'windspeed']]
data.head(3)

# Create lagged features and rolling windows features from exogenous variables
# ==============================================================================
wf_transformer = WindowFeatures(
                    variables           = ["temp", "windspeed"],
                    window            = ["24h"],
                    functions          = ["mean"],
                    freq                  = "h",
                    missing_values = "ignore",
                    drop_na           = False,
                )
lag_transformer = LagFeatures(
                    variables = ["temp", "windspeed"],
                    periods   = [1, 2, 3, 4, 5],
                )
exog_transformer = make_pipeline(
                        wf_transformer,
                        lag_transformer
                    )
exog_transformer.fit_transform(data)
```

This approach is effective for automating the creation of lagged and window-based features from exogenous variables. However, it is not compatible with the current implementation of the transform_exog argument in skforecast. The issue arises in the predict method, where the exog argument expects a series or dataframe starting immediately after last_window. Consequently, there is no data available to compute lagged values at this stage, and the exogenous values for the first predictions would be missing.

**Reply by JDA75** [reply]
Hi, do you have idea on when this feature would be working inside the transformer_exog ?

**Comment by JoaquinAmatRodrigo** [comment]
Hi @JDA75,
It would be quite challenging to support this kind of transformation directly via transformer_exog, mainly because there isn't a unified or reliable way to determine the window size or required historical context from the transformer itself. This makes it hard to ensure compatibility with how exog is passed at prediction time.

Do you see a significant disadvantage to creating the lagged and window-based features outside the forecaster beforehand? This approach provides more control and transparency, and while it adds an extra step, it tends to be more robust and flexible given the current design constraints.
